File: WOT-iOS/WOT-iOS/ui/Root/WOTMenuViewController/WOTMenuViewController.m
//
//  WOTMenuViewController.m
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 6/3/15.
//  Copyright (c) 2015 Pavel Yeshchyk. All rights reserved.
//

#import "WOTMenuViewController.h"
#import "WOTMenuTableViewCell.h"
#import "UserSession.h"

#import "WOTSessionDataProvider.h"
#import "WOTMenuDatasource.h"

@interface WOTMenuViewController () <UITableViewDataSource, UITableViewDelegate, WOTMenuDatasourceDelegate>

@property (nonatomic, weak) IBOutlet UITableView *tableView;
@property (nonatomic, assign) NSInteger selectedIndex;
@property (nonatomic, strong) WOTMenuDatasource *menuDatasource;

@end

@implementation WOTMenuViewController
@synthesize selectedMenuItemClass;
@synthesize delegate;

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {

    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self){
        
        self.menuDatasource = [[WOTMenuDatasource alloc] init];
        self.menuDatasource.delegate = self;
        [self.menuDatasource rebuild];
        self.selectedIndex = 0;
    }
    return self;
}

- (void)dealloc {
    
    self.menuDatasource.delegate = nil;
}

- (Class )selectedMenuItemClass {
    
    WOTMenuItem *item = [self.menuDatasource objectAtIndex:self.selectedIndex];
    return item.controllerClass;
}

- (UIImage *)selectedMenuItemImage {
    
    WOTMenuItem *item = [self.menuDatasource objectAtIndex:self.selectedIndex];
    return item.icon;
}

- (NSString *)selectedMenuItemTitle {
    
    WOTMenuItem *item = [self.menuDatasource objectAtIndex:self.selectedIndex];
    return item.controllerTitle;
}

- (void)viewDidLoad {

    [super viewDidLoad];
    
    self.tableView.backgroundColor = [UIColor colorWithRed:45.0f/255.0f green:45.0f/255.0f blue:45.0f/255.0f alpha:1.0f];
    self.view.backgroundColor = [UIColor colorWithRed:45.0f/255.0f green:45.0f/255.0f blue:45.0f/255.0f alpha:1.0f];
    
    [self.tableView registerNib:[UINib nibWithNibName:NSStringFromClass([WOTMenuTableViewCell class]) bundle:nil] forCellReuseIdentifier:NSStringFromClass([WOTMenuTableViewCell class])];

    [self redrawNavigationBar];
    
}


- (void)setSelectedIndex:(NSInteger)selectedIndex {

    _selectedIndex = selectedIndex;
    [self.delegate menu:self didSelectControllerClass:self.selectedMenuItemClass title:self.selectedMenuItemTitle image:self.selectedMenuItemImage];
}

#pragma mark - UITableViewDatasource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {

    return [self.menuDatasource objectsCount];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    WOTMenuTableViewCell *result = [tableView dequeueReusableCellWithIdentifier:NSStringFromClass([WOTMenuTableViewCell class]) forIndexPath:indexPath];
    WOTMenuItem *menuItem = [self.menuDatasource objectAtIndex:indexPath.row];
    result.cellTitle = menuItem.controllerTitle;
    result.cellImage = menuItem.icon;
    return result;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {

    [tableView deselectRowAtIndexPath:indexPath animated:NO];
    
    self.selectedIndex = indexPath.row;
}

#pragma mark - WOTMenuDatasourceDelegate
- (void)hasUpdatedData:(WOTMenuDatasource *)datasource {
    
    [self redrawNavigationBar];
    self.selectedIndex = 0;

    [self.tableView reloadData];
}

#pragma mark - private

- (void)redrawNavigationBar {

    [self.navigationController.navigationBar setDarkStyle];

    UIImage *image = [UIImage imageNamed:WOTString(WOT_IMAGE_MENU_ICON)];
    UIBarButtonItem *backButtonItem = [UIBarButtonItem barButtonItemForImage:image text:nil eventBlock:^(id sender) {
        
        [self.delegate loginPressedOnMenu:self];
    }];
    [self.navigationItem setLeftBarButtonItems:@[backButtonItem]];
    [self.navigationItem setTitle:self.delegate.currentUserName];
}


@end

File: WOT-iOS/WOT-iOS/ui/Root/WOTMenuViewController/WOTMenuViewController.swift
//
//  WOTMenuViewControllerSwift.swift
//  WOT-iOS
//
//  Created on 8/14/18.
//  Copyright © 2018. All rights reserved.
//

import Foundation

@objc
protocol WOTMenuDelegate: NSObjectProtocol {
    var currentUserName: String { get }
    func menu(_ menu: WOTMenuProtocol, didSelectControllerClass controllerClass: AnyClass, title: String, image: UIImage)
    func loginPressedOnMenu(_ menu: WOTMenuProtocol)

}

@objc
protocol WOTMenuProtocol: NSObjectProtocol {
    var delegate: WOTMenuDelegate? { get set }
    var selectedMenuItemClass: AnyClass { get }
    var selectedMenuItemTitle: String { get }
    var selectedMenuItemImage: UIImage { get }
}

@objc(WOTMenuViewController)

class WOTMenuViewController: UIViewController, WOTMenuProtocol {

    @IBOutlet var tableView: UITableView?
    var selectedIndex: NSInteger = 0 {
        didSet {
            self.delegate?.menu(self, didSelectControllerClass: self.selectedMenuItemClass, title: self.selectedMenuItemTitle, image: self.selectedMenuItemImage)
        }
    }
    var menuDatasource: WOTMenuDatasourceProtocol?

    weak var delegate: WOTMenuDelegate?
    var selectedMenuItemClass: AnyClass {
        let item = self.menuDatasource?.object(at: self.selectedIndex)
        return item?.controllerClass ?? WOTMenuViewController.self
    }

    var selectedMenuItemTitle: String {
        let item = self.menuDatasource?.object(at: self.selectedIndex)
        return item?.controllerTitle ?? ""
    }

    var selectedMenuItemImage: UIImage {
        let item = self.menuDatasource?.object(at: self.selectedIndex)
        return item?.icon ?? UIImage()
    }

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        self.menuDatasource = WOTMenuDatasource()
        self.menuDatasource?.delegate = self
        self.menuDatasource?.rebuild()
        self.selectedIndex = 0

    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        //        self.tableView.backgroundColor = WOT_COLOR_DARK_VIEW_BACKGROUND;
        //        self.view.backgroundColor = WOT_COLOR_DARK_VIEW_BACKGROUND;

        let nib = UINib(nibName: String(describing: WOTMenuTableViewCell.self), bundle: nil)
        self.tableView?.register(nib, forCellReuseIdentifier: String(describing: WOTMenuTableViewCell.self))

        self.redrawNavigationBar()

    }

    fileprivate func redrawNavigationBar() {
        self.navigationController?.navigationBar .setDarkStyle()
        let image = UIImage(named: L10n.wotImageUser)
        let backButton = UIBarButtonItem(image: image, style: UIBarButtonItem.Style.done, target: self, action: #selector(WOTMenuViewController.loginPressed(_ :)))
        self.navigationItem.leftBarButtonItems = [backButton]
        self.navigationItem.title = self.delegate?.currentUserName
    }

    @objc func loginPressed(_ obj: AnyObject?) {
        self.delegate?.loginPressedOnMenu(self)
    }
}

extension WOTMenuViewController: WOTMenuDatasourceDelegate {
    func hasUpdatedData(_ datasource: WOTMenuDatasourceProtocol) {
        self.redrawNavigationBar()
        self.selectedIndex = 0
        self.tableView?.reloadData()
    }
}

extension WOTMenuViewController: UITableViewDataSource, UITableViewDelegate {

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.menuDatasource?.objectsCount() ?? 0
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard  let result = tableView.dequeueReusableCell(withIdentifier: String(describing: WOTMenuTableViewCell.self), for: indexPath) as? WOTMenuTableViewCell else {
            return UITableViewCell()
        }
        if let menuItem = self.menuDatasource?.object(at: indexPath.row) {
            result.cellTitle = menuItem.controllerTitle
            result.cellImage = menuItem.icon
        }
        return result
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: false)
        self.selectedIndex = indexPath.row
    }

}
